minimint-api/src/db/mod.rs

```diff
@@ -124,8 +158,129 @@ impl<'a> dyn Database + 'a {
         })
     }
 
-    pub fn apply_batch(&self, batch: DbBatch) -> Result<()> {
-        self.raw_apply_batch(batch)
+    pub fn transaction<'a>(&'a self) -> Result<Box<dyn Transaction<'a>>> {
+        self.transaction()
+    }
+}
+
+impl<'a> dyn Transaction<'a> {
```

Comment:
Perhaps a naive rust question: what is the purpose of this implementation? Is it essentially a wrapper around the core rocksdb bindings?

Reply:
It is a intermediary between the core bindings it helps ensure that all users of the API are supplying compliant data-types and it also handles pointing to the bindings while also handling encoding and decoding of data from bytes.

Reply:
My understanding is that this `impl` block this is done to introduce bunch of methods for `Transaction` that are not part of the `trait Transaction` itself (so implementations don't get to potentially implemented them - otherwise default implementation would be good enough).

However I'm used to this being implemented generically via extension trait + blanket implementation, [like this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e80776695b2536b57fbd663098ca4386):

```rust
trait Foo {
    fn make_foo(&self);
}

trait FooExt: Foo {
    fn make_double_foo(&self) {
        self.make_foo();
        self.make_foo();
    }
}

impl<T> FooExt for T where T: Foo + ?Sized {}

struct Bar;

impl Foo for Bar {
    fn make_foo(&self) {
        println!("foo");
    }
}

fn main() {
    let f1 = Bar;
    let f2 = Box::new(Bar);
    let f3 = Box::new(Bar) as Box<dyn Foo>;

    f1.make_double_foo();
    f2.make_double_foo();
    f3.make_double_foo();
}
```

@elsirion Since I've seen `impl<'a> dyn SomeTrait<'a> {` couple of times around the codebase - are there any reasons to prefer non-generic impl?

Reply:
BTW. Here it would maybe make sure to name them `trait RawTransaction` and `trait Transaction : RawTransaction`, clarifying the difference via name.